### PR TITLE
Search stylin'

### PR DESCRIPTION
**Pull request by hamburge**
Issues #1816, #1807, and also:
- Fix for if you had a filter set, the autocomplete suggestion would not always be lined up depending on your screen resolution.
- Fix for if there were no filter suggestions, there was an extra line on the top of the drop down.
- When you click on the dialog and the input loses focus, don't show the autofill suggestion
- Removes code that supported displaying multiple svgs for filters since it is not necessary.

### CODE REVIEW

File: src/js/jsx/Search.jsx [hunk 1]
```diff
@@ -86,6 +86,7 @@ define(function (require, exports, module) {
                             searchID={SEARCH_BAR_DIALOG_ID}
                             dismissDialog={this._closeSearchBar}
                             executeOption={this._handleOption}
+                            maxOptions={14}
```
Comment (iwehrman) [inline src/js/jsx/Search.jsx]: It will be easier to change this later if it's broken out into a `CONSTANT` with JSDoc.

Comment (iwehrman) [inline src/js/jsx/Search.jsx]: Or, just set the default the way you want it for `Search` in `SearchBar` below and omit this entirely. (Future clients of `SearchBar` can change the default as they see fit, but we might as well make the default correct for the first client.)

File: src/js/jsx/shared/Datalist.jsx [hunk 2]
```diff
@@ -451,6 +475,15 @@ define(function (require, exports, module) {
             }
         },
 
+        removeAutofillSuggestion: function () {
```
Comment (iwehrman) [inline src/js/jsx/shared/Datalist.jsx]: JSDoc needed.

File: src/js/jsx/shared/Datalist.jsx [hunk 3]
```diff
@@ -451,6 +475,15 @@ define(function (require, exports, module) {
...
+            if (this.props.useAutofill) {
+                this.setState({
+                    suggestTitle: ""
+                });
+                this.forceUpdate();
```
Comment (iwehrman) [inline src/js/jsx/shared/Datalist.jsx]: The `setState` call above will cause `render` to be run assuming the state passes through `shouldComponentUpdate`. Why is the extra `forceUpdate` call necessary?

File: src/js/jsx/shared/Datalist.jsx [hunk 4]
```diff
@@ -527,7 +560,8 @@ define(function (require, exports, module) {
             }
             
             var autocomplete,
-                hiddenTI;
+                hiddenTI,
+                hiddenSVG;
  
             if (this.props.useAutofill) {
                 hiddenTI = (
```
Comment (iwehrman) [inline src/js/jsx/shared/Datalist.jsx]: Can you leave a few `// comments` here that describe the point of these hidden elements?

File: src/style/search/search-bar.less [hunk 5]
```diff
@@ -67,10 +71,17 @@
 .search-bar__dialog .drop-down .hiddeninput {
     display: inline-block;
     position: absolute;
-    left: 3.2rem;
+    margin-left: .1rem;
     visibility: hidden;
     white-space: pre;
 }
+.search-bar__dialog .drop-down .hiddenSVG {
```
Comment (iwehrman) [inline src/style/search/search-bar.less]: The naming on some of these classes is not consistent with the rules in our style guide here: https://github.com/adobe-photoshop/spaces-design/wiki/Coding-Conventions#naming-and-syntax

**Comment by iwehrman** [comment]
A few minor requests. Back to @hamburge for polish.

File: src/js/jsx/shared/Datalist.jsx (hunk 3 above)
Comment (hamburge) [inline src/js/jsx/shared/Datalist.jsx]: Otherwise in `shouldComponentUpdate`, it sets `suggestTitle` to a value other than `""` again before rendering.

Comment (iwehrman) [inline src/js/jsx/shared/Datalist.jsx]: This is a bad code smell. `shouldComponentUpdate` really should not have side effects. Why is that necessary? What are other ways that we could get rid of this extra call? If I'm understanding this, I think the proposed control flow is too messy to commit.

Comment (volfied) [inline src/js/jsx/shared/Datalist.jsx]: Yeah this is bad... SCU should not ever update anything. The render method should see the conditions you're checking for in SCU and render correctly if the condition is being met.

Comment (hamburge) [inline src/js/jsx/shared/Datalist.jsx]: I moved the call I was making in SCU to `inputChange`, which I think addresses both of the issues.

**Comment by hamburge** [comment]
@iwehrman I cleaned it up based on your and @volfied's comments so now there shouldn't be side effects from `shouldComponentUpdate`

File: src/js/util/svg.js [hunk 6]
```diff
@@ -54,40 +54,33 @@ define(function (require, exports) {
...
             if (kind === "ARTBOARD") {
                 iconID += "artboard";
             } else if (kind === "BACKGROUND") {
                 iconID += layerLib.layerKinds.PIXEL;
             } else if (kind === "SMARTOBJECT" && isLinked) {
                 iconID += layerLib.layerKinds.SMARTOBJECT + "-linked";
-            } else if (kind !== "LAYER") {
+            } else if (kind.indexOf("LAYER") === -1) { // check if kind is "ALL_LAYER" or "CURRENT_LAYER"
                 iconID += layerLib.layerKinds[kind.toUpperCase().replace(" ", "")];
             }
```
Comment (volfied) [inline src/js/util/svg.js]: What about the cases where none of these conditions match, is `"layer-"` an acceptable iconID?

File: src/js/jsx/shared/Datalist.jsx [hunk 7]
```diff
@@ -519,20 +552,35 @@ define(function (require, exports, module) {
...
             }
             
-            var autocomplete,
-                hiddenTI;
+            // Use hidden text input and hidden svg to find position of suggestion by updating their values and 
+            // getting width before calling render. Without using the hidden elements, in certain cases the 
+            // calculated position of suggestion would be one keystroke behind. 
+
+            // Hidden text input gets the value of the search input before rendering.
+            var hiddenTI,
```
Comment (volfied) [inline src/js/jsx/shared/Datalist.jsx]: I believe React prefers components to be defined, or `null`. Can you null initialize them here?

File: src/js/jsx/shared/Datalist.jsx [hunk 8]
```diff
@@ -580,22 +628,24 @@ define(function (require, exports, module) {
...
-
-            if (this.props.filterIcons) {
-                _.forEach(this.props.filterIcons, function (icon, index) {
-                    svg.push((<SVGIcon
-                            className="datalist__svg"
-                            CSSID={icon}
-                            key={index}
-                            viewbox="0 0 24 24"/>));
-                });
+                svg;
```
Comment (volfied) [inline src/js/jsx/shared/Datalist.jsx]: diitto

File: src/js/jsx/shared/Datalist.jsx [hunk 9]
```diff
@@ -393,33 +389,58 @@ define(function (require, exports, module) {
...
+         * @private
+         * @param {string} value The current value of the input
+         * @param {string} icon The SVG icon for the input
+         */
+        _updateAutofillPosition: function (value, icon) {
+            // Base position off of hidden element widths because Datalist sub-components
+            // haven't rendered yet using the new value, and we don't want the suggestion to move
+            // after rendering
+
+            var hiddenInput = React.findDOMNode(this.refs.hiddenTextInput);
```
Comment (volfied) [inline src/js/jsx/shared/Datalist.jsx]: https://facebook.github.io/react/tips/dangerously-set-inner-html.html

Given this link, can you justify using innerHTML here?

**Comment by volfied** [comment]
This code is diverging from React. I have couple nits, and one big problem.

File: src/js/jsx/shared/Datalist.jsx (hunk 9 above)
Comment (hamburge) [inline src/js/jsx/shared/Datalist.jsx]: Wow so scary. For some reason I originally thought that moving the autofill suggestion in `componentDidUpdate` would be bad but it looks like that's possible, so I can avoid using `innerHTML` :grin:

File: src/js/util/svg.js (hunk 6 above)
Comment (hamburge) [inline src/js/util/svg.js]: That shouldn't ever happen. If the filter just contains "ALL_LAYER" or "CURRENT_LAYER" then it returns above. And otherwise the options are all in layerKinds, so they'll give the correct value.

I realize the logic to get the icons is very layer-specific right now, but once I get the other SVGs from Seth, I'll work on making it all more generic.

**Comment by hamburge** [comment]
@volfied Ok reworked it so that it is more React-friendly, I think.
